### ISSUE TITLE
Create psammead bumper bot (strictly for psammead packages)

### ISSUE THREAD

**Issue by dr3** [BBC-archive/psammead#888]
**Is your feature request related to a problem? Please describe.**
We are falling behind on psammead bumps, we should create a bot to bump dependancies for us. We tried dependabot but it opened 100PRs with single bumps per PR which isn't ideal

**Describe the solution you'd like**
When a psammead PR is merged & published
 -> For each newly published package, bump all dependants (patch, minor, major) in psammead
 -> Run `npm run updateSnapshots`
 ->  -> Make a single draft PR with a summary of the changes.

The changelog and version bumping need to be an automated part of this solution. This would probably best be done as a separate PR that precedes this primary implementation.

**Testing notes**
No testing required

**Additional context**
Please write in JS or bash. Not enough devs here write python so its unmaintainable.
Should be public and OSS, ideally in this repo.

**Comment by dr3** [comment]
Please discuss this across pods before starting the work.

**Comment by dr3** [comment]
I dont think this should bump all package deps by default. I should have the ability for that, but the primary function should be bumping psammead packages within itself.

On PR merge
1) Get all JUST published packages
2) For each newly published package bump its version in every dependant psammead package.
3) run `npm i` at every dependant level
4) Add to changelog to say what has been changed at every dependant level
5) Run `npm run updateSnapshots` at top level
6) Create a PR with all new changes.

**Comment by chris-hinds** [comment]
Why limit to just psammead? I would make this a little more generic, so it would with Simorgh too.

**Comment by dr3** [comment]
@hindsc52 What did you have in mind? I personally dont think we should use a generic bot that just bumps everything at once as the PRs it creates are basically untestable as they have a crazy scope.

**Comment by chris-hinds** [comment]
@dr3 not bumping everything at once, but just making sure this isn't tied into Psammead alone. It would be nice if this can be passed config for when it runs, or how many PR's it can raise per day etc.

Just would be nice to have one single bot/service for update handling.

**Comment by dr3** [comment]
@hindsc52 Can you please write out a logic flow like https://github.com/bbc/psammead/issues/888#issuecomment-509223636 please, I dont follow what you think it should do

**Comment by chris-hinds** [comment]
So i think we are thinking of two different services really. I am more thinking of a dependant style service but with more customisations.

Correct me if I am wrong but what you are suggesting would effectively double the psammead PR's right? as most PR's will cause a new version bump PR once merged.

As in Dev merges, this services then raises another PR to update other packages.

**Comment by dr3** [comment]
@hindsc52 yeah thats what im thinking of, i think its a nice balance between 1 PR per every dependant package (like dependabot) or 1 PR every X time for everything. This way, each following PR, even if its 7 packages that are bumped, all have related changes, making testing alot easier

**Comment by pjlee11** [comment]
Unless I'm misunderstanding ['On PR merge' in this comment](https://github.com/bbc/psammead/issues/888#issuecomment-509223636) isn't feasible on nested bumps. Take for example: 

```
package-a
- package-b
-- package-c
```

Given the `package-c` is updated the bot will need to bump `package-b` with the new install of `package-c`. Then only after `package-b` is deployed/published can the bot bump `package-a` with the new install of `package-b` & `package-c`. 

If this isn't handled in 2 PRs the bot would have to chain deploys and could mean that failures happen after some successful deploys take place.

I'm not against this idea but want to ensure everyone is aware that this would be necessary.

**Comment by pjlee11** [comment]
Separately I have reservations on a bot bumping `major` versions. This is because new props should be passed and if we are bumping to a new major version it is likely a change will be needed, therefore we will manually want to see through. There is scope for automating this with something that is only ever a `devDep` so maybe it `major` bumps could be behind a flag with when triggering the bot

**Comment by dr3** [comment]
@pjlee11 isn't this already a psammead limitation? You cant install a package before it merged, this just does it automatically

**Comment by pjlee11** [comment]
It's a limitation of our psammead deploy process rather than psammead itself. We could easily have PRs changing multiple levels of nested packages on the condition we are happy to automate the following in the deployment pipeline: 
- the ordering of how the dependencies are deployed 
- the installation of new deps without review
- the update of snapshots without review

The changelog, version number and readme could be updated in the the original PR

Edit: changed from `without PR and review` to `without review` as the PR would be wanted so the changes are in github before then being deployed to npm. But the PR would be generated and merged with review therefore triggering the next deploy

Also caveating this as it could break if we don't get the process right. It also pretty much ignores what lerna currently handles for us

**Comment by dr3** [comment]
Hmm I personally dont think we should change that, espesh not atm. The restriction of only making 1 level of dependancy changes isnt that bad, it helps keep PRs isolated to 1 task

**Comment by dr3** [comment]
The easiest way I see to integrate this with our existing tools. This is all in the existing jenkins pipeline. All scripts will be written in JS as we are a JS team :)

1) When psammead publish script publishes a package, add its name on a new line in a txt file. This will build a list of newly published packages

2) In a new, separate, script read off this file.
3) For each line (package name), use lerna to bump any places its used in psammead
4) Bump any packages with newly update dependancies
5) Add an entry to its changelog saying what we bumped
6) run `npm run updateSnapshots`
7) Using the jenkins github token, create a branch with the new changes, push it.
8) Use the github API to make a PR with this branch

A developer would then see the new PR via slack, see if jenkins is failing or not, see changes it pulls in and sanity check, fix any major changes & its ready for test

Extra things we can do for fun
1) In the generated PR, put what has been bumped, and put a copy of the changes made in the description. These can be got from the changelogs of the newly generated PRs

**Comment by dr3** [comment]
Pseudo code :)

// Existing publish script
```
if (successful publish 'packageName') {
  let file = getTxtFile('publishedPackages.txt')
  file += '/n' + packageName
  writeTxtFile('publishedPackages.txt')
}
```
// New script
```
const allPackages = lernaGetAllPackages();
const publishedPackages = getTxtFile('publishedPackages.txt').split('/n');
const changedPackages = [];

publishedPackage.forEach(publishedPackageName => {
  allPackages.forEach(packageName => {
    if(packageName is dependant on publishedPackageName) {
      update({ dependant: packageName, dependancy: publishedPackageName });
      changedPackages.push(packageName);
    }
  })
})

changedPackages = removeDuplicates(changedPackages);

changedPackages.forEach(package => {
  bumpVersionPatch(package);
  updateChangelog(package);
  npmInstall(package);
});

updateSnapshots();

createBranch();
commitChanges();
pushBranch();
makePr();
```

**Comment by dr3** [comment]
Meeting decided - Dont automatically update snapshots

**Comment by dr3** [comment]
Todo: Investigate how to find what uses a dependancy.
Todo: Investigate how to make a branch & Github PR (is there a package to do this?)
Todo, exclude psammead packages from live

Nice to have: Investigate assigning new PR to original author

**Comment by thekp** [comment]
bumping versions and update changelog POC: https://github.com/bbc/psammead/pull/890
